modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/ExpressionFactoryImpl.java

```diff
@@ -237,7 +239,8 @@ else if (c2 != HIGHEST_VALUE)
                 }
 
                 // If compared rows contain NULLs, they shouldn't be treated as equals, since NULL <> NULL in SQL.
-                return hasNulls ? 1 : 0;
+                // Expect for cases with IS DISTINCT / IS NOT DISTINCT.
+                return hasNulls ? nullComparison : 0;
```

Comment:
Let's use just `return hasNulls && !nullsEqual ? 1 : 0;`
`left.get(0).nullDirection.nullComparison` - null comparison for first field, I think it's not better than constant.

Reply:
Nulls comparison must be the same to the `right`. It is checked above:

```
        // Check that collations is correct.
        for (int i = 0; i < left.size(); i++) {
            if (left.get(i).nullDirection.nullComparison != right.get(i).nullDirection.nullComparison)
                throw new IllegalArgumentException("Can't be compared: left=" + left.get(i) + ", right=" + right.get(i));

            if (left.get(i).direction != right.get(i).direction)
                throw new IllegalArgumentException("Can't be compared: left=" + left.get(i) + ", right=" + right.get(i));
        }
```

While returning `1` seems to work, taking a real comparison value covers theoretical common case. Let's keep it. WDYT?

Reply:
But you always return nullComparison for field 0, while nulls can be on another field. So it's not better than just return 1, which means that it's not equals.